src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/CodeActions/Razor/UnboundDirectiveAttributeAddUsingCodeActionProvider.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.Threading;
+using Microsoft.CodeAnalysis.Razor.CodeActions.Models;
+using Microsoft.CodeAnalysis.Razor.CodeActions.Razor;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+
+namespace Microsoft.CodeAnalysis.Razor.CodeActions;
+
+internal class UnboundDirectiveAttributeAddUsingCodeActionProvider : IRazorCodeActionProvider
+{
+    public Task<ImmutableArray<RazorVSInternalCodeAction>> ProvideAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+    {
+        if (context.HasSelection)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        // Only work in component files
+        if (!FileKinds.IsComponent(context.CodeDocument.FileKind))
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        if (!context.CodeDocument.TryGetSyntaxRoot(out var syntaxRoot))
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        // Find the node at the cursor position
+        var owner = syntaxRoot.FindInnermostNode(context.StartAbsoluteIndex, includeWhitespace: false);
+        if (owner is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        // Find a regular markup attribute (not a tag helper attribute) that starts with '@'
+        // Unbound directive attributes are just regular attributes that happen to start with '@'
+        var attributeBlock = owner.FirstAncestorOrSelf<MarkupAttributeBlockSyntax>();
+        if (attributeBlock is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        // Make sure the cursor is actually on the name part, since the attribute block is the whole attribute, including
+        // value and even some whitespace
+        if (!attributeBlock.Name.Span.Contains(context.StartAbsoluteIndex))
```

Comment:
>Contains [](http://example.com/codeflow?start=38&length=8)

@davidwengier Do we normally not include the end position in these tests?

Reply:
The first guard clause in this checks for a selection, and doesn't offer the action if there is one, so end and start will always be the same. I don't know if that is best practice for code actions, but I think most, if not all, of the Razor ones do it.

Reply:
Sorry, that wasn't clear of me, I meant the end position of the Name.span. This will trigger if the caret is here

<a []@foo="test" />

but not here

<a @foo[] = "test" />